build_runner/lib/src/generate/watch_impl.dart

```diff
@@ -73,54 +73,81 @@ class WatchImpl implements BuildState {
   /// Runs a build any time relevant files change.
   ///
   /// Only one build will run at a time, and changes are batched.
-  Stream<BuildResult> _run(BuildOptions options, List<BuildAction> buildActions,
-      Future until) async* {
-    var fatalBuild = new Completer();
-    var firstBuild = new Completer<BuildResult>();
-    currentBuild = firstBuild.future;
-    var changes = new PackageGraphWatcher(packageGraph,
-        logger: _logger,
-        watch: (node) => new PackageNodeWatcher(node, watch: (path) {
-              return _directoryWatcherFactory(path);
-            })).watch();
-
-    var buildDefinition = await BuildDefinition.load(options, buildActions);
-    _readerCompleter.complete(buildDefinition.reader);
-    _assetGraph = buildDefinition.assetGraph;
-
-    var build = await BuildImpl.create(buildDefinition, buildActions,
-        onDelete: _expectedDeletes.add);
-    yield build.firstBuild;
-    firstBuild.complete(build.firstBuild);
+  ///
+  /// File watchers are scheduled synchronously.
+  Stream<BuildResult> _run(
+      BuildOptions options, List<BuildAction> buildActions, Future until) {
+    var fatalBuildCompleter = new Completer();
+    var firstBuildCompleter = new Completer<BuildResult>();
+    currentBuild = firstBuildCompleter.future;
+    var controller = new StreamController<BuildResult>();
+
+    // We need this inside `doBuild`, but we won't actually create it until
```

Comment:
FWIW I think this would be easier to follow with the old structure and a `Future<Stream>` returned from `watch` but I don't have strong feelings about it.

Reply:
We would still need to define the `doBuild` function before a `BuildImpl` actually exists though, unless we jumped through some major hoops. The issue is that `BuildImpl.create()` actually schedules the first build, and even if it didn't we still want to start the watchers before we create the `BuildDefinition` to eliminate race conditions. The only thing it would allow us to do is eliminate the async closure that actually creates the build.